flatdata-py/flatdata/lib/archive_builder.py

```diff
@@ -0,0 +1,290 @@
+        '''
+        Creates and writes instance type resource
+
+        :param storage(object): handles storage and writing to disc
+        :param name(str): instance name
+        :param value(dict): instance object replicates struct
+        '''
+        initializer = self._RESOURCES[name].initializer
+        ArchiveBuilder.validate_structure_fields(name, value, initializer)
+
+        bout = bytearray(initializer._SIZE_IN_BYTES)
+        for (key, field) in initializer._FIELDS.items():
+            write_value(bout, field.offset, field.width,
+                        field.is_signed, value[key])
+
+        storage.write(bout)
+
+    def set_vector(self, storage, name, vector):
+        '''
+        Creates and writes vector resource
+
+        :param storage(object): handles storage and writing to disc
+        :param name(str): resource name
+        :param vector(list): vector, provided as list of dict ie [{},{}]
+        '''
+        initializer = self._RESOURCES[name].initializer
+        for value in vector:
+            ArchiveBuilder.validate_structure_fields(name, value, initializer)
+        for value in vector:
+            bout = bytearray(initializer._SIZE_IN_BYTES)
+            for (key, field) in initializer._FIELDS.items():
+                write_value(bout, field.offset, field.width,
+                            field.is_signed, value[key])
+            storage.write(bout)
+
+    def set_multivector(self, storage, name, value):
+        '''
+        Creates and writes multivector resource
+
+        :param storage(object): handles storage and writing to disc
+        :param name(str): resource name
+        :param value(list): mulitvector, provided as list of list of dict ie [[{},{}],[]]
+        '''
+        initializer_list = self._RESOURCES[name].initializer
+
+        def valid_structure_name(_obj):
+            return _obj['name'] in [_initializer._NAME for _initializer in initializer_list[1:]]
+
+        def validate_fields(_obj):
+            matched_obj_list = [
+                _initializer for _initializer in initializer_list[1:] if _initializer._NAME == _obj['name']]
+            if len(matched_obj_list) == 1:
+                ArchiveBuilder.validate_structure_fields(
+                    name, _obj['attributes'], matched_obj_list[0])
+
+        for sub_list in value:
+            for obj in sub_list:
+                if not valid_structure_name(obj):
+                    raise UnknownStructureError(obj['name'])
+                validate_fields(obj)
+
+        index_data_points = []
+        data_point = 0
+        data_size = 0
+
+        for sub_list in value:
+            # index_writer.add(data_point)
+            index_data_points.append(data_point)
+            if len(sub_list) > 0:
+                for obj in sub_list:
+                    # find out correct initializer
+                    type_index, matched_initializer = [(index, element) for index, element in enumerate(
```

Comment:
Do you think it is worth just creating a small map in the beginning of the function?

Reply:
Yes, it can be done.